File: src/Ipe/Prelude/Json/Encode.js
/**
 * @typedef {import("../Json").JsonValue} JsonValue
 */

/**
 *
 * @param {string} input The string to encode
 * @returns {string} The encoded string
 */
const string = (input) => input;

/**
 *
 * @param {boolean} input The boolean to encode
 * @returns {boolean} The encoded boolean
 */
const bool = (input) => input;

/**
 *
 * @param {number} input The number to encode
 * @returns {number} The encoded number
 */
const number = (input) => input;

/**
 *
 * @returns {null} The encoded null
 */
const nothing = () => null;

/**
 * @template T
 * @param {(input: T) => JsonValue} encodeItem
 * @param {T[]} arr
 * @returns {JsonValue}
 */
const list = (encodeItem, arr) => arr.map(encodeItem);

/**
 * @param {Map<string, JsonValue>} input
 * @returns {JsonValue}
 */
const object = (input) => {
  const result = {};
  for (const [key, value] of input) {
    result[key] = value;
  }
  return result;
};

export default {
  string,
  bool,
  number,
  nothing,
  list,
  object,
};

File: src/Ipe/Prelude/Json/Encode.hs
{-# LANGUAGE OverloadedStrings #-}

module Ipe.Prelude.Json.Encode (moduleTypes) where

import qualified Data.Map as Map
import Data.Text (Text)
import Ipe.TypeChecker.Utils (Type (..))

moduleTypes :: Map.Map Text Type
moduleTypes =
  Map.fromList
    [ ("string", TFun TStr jsonValue),
      ("number", TFun TNum jsonValue),
      ("nothing", TFun (TRec []) jsonValue),
      ( "list",
        TFun
          (TFun (TVar "input") jsonValue)
          ( TFun
              (TList (TVar "input"))
              jsonValue
          )
      ),
      ( "object",
        TFun dictType jsonValue
      )
    ]

jsonValue :: Type
jsonValue = TCustom "Json.Value" [] []

dictType :: Type
dictType = TCustom "Dict.Dict" [TStr, jsonValue] []
